File: src/main/java/origami/filters/HueSaturationValue.java
package origami.filters;

import org.opencv.core.Mat;
import origami.Filter;
import static org.opencv.core.Core.multiply;
import static org.opencv.core.CvType.CV_64F;
import static org.opencv.imgproc.Imgproc.*;

public class HueSaturationValue {

    static Mat filter(Mat marcel, int cs1, int cs2, double... values) {
        Mat m = new Mat(1, 3, CV_64F);
        m.put(0, 0, values);
        if (cs1 != -1)
            cvtColor(marcel, marcel, cs1);
        multiply(marcel, m, marcel);
        if (cs2 != -1)
            cvtColor(marcel, marcel, cs2);
        return marcel;
    }

    public static class Pink implements Filter {

        @Override
        public Mat apply(Mat in) {
            filter(in, COLOR_BGR2HSV, COLOR_HSV2BGR, 0.2, 1.0, 1.0);
            return in;
        }
    }

    public static class Nashville implements Filter {
        @Override
        public Mat apply(Mat marcel) {
            filter(marcel, COLOR_BGR2HSV, COLOR_HSV2BGR, 1.0, 1.5, 1.0);
            marcel.convertTo(marcel, -1, 1.5, -30);
            return marcel;
        }
    }

    public static class Gotham implements Filter {
        @Override
        public Mat apply(Mat marcel) {
            filter(marcel, COLOR_BGR2HSV, COLOR_HSV2BGR, 1.0, 0.2, 1.0);
            marcel.convertTo(marcel, -1, 1.3, -20);
            return marcel;
        }
    }

    public static class Lomo implements Filter {

        @Override
        public Mat apply(Mat in) {
            filter(in, -1, -1, 1.0, 1.33, 1.33);
            return in;
        }
    }

}

File: src/main/java/origami/filters/HueSaturationValue.kt
package origami.filters

import org.opencv.core.Core
import org.opencv.core.CvType
import org.opencv.core.Mat
import org.opencv.imgproc.Imgproc
import origami.Filter

object HueSaturationValue {
    fun filter(marcel: Mat, cs1: Int, cs2: Int, vararg values: Double): Mat {
        val m = Mat(1, 3, CvType.CV_64F)
        m.put(0, 0, *values)
        if (cs1 != -1) Imgproc.cvtColor(marcel, marcel, cs1)
        Core.multiply(marcel, m, marcel)
        if (cs2 != -1) Imgproc.cvtColor(marcel, marcel, cs2)
        return marcel
    }

    class Pink : Filter {
        override fun apply(`in`: Mat): Mat {
            filter(`in`, Imgproc.COLOR_BGR2HSV, Imgproc.COLOR_HSV2BGR, 0.2, 1.0, 1.0)
            return `in`
        }
    }

    class Nashville : Filter {
        override fun apply(marcel: Mat): Mat {
            filter(marcel, Imgproc.COLOR_BGR2HSV, Imgproc.COLOR_HSV2BGR, 1.0, 1.5, 1.0)
            marcel.convertTo(marcel, -1, 1.5, -30.0)
            return marcel
        }
    }

    class Gotham : Filter {
        override fun apply(marcel: Mat): Mat {
            filter(marcel, Imgproc.COLOR_BGR2HSV, Imgproc.COLOR_HSV2BGR, 1.0, 0.2, 1.0)
            marcel.convertTo(marcel, -1, 1.3, -20.0)
            return marcel
        }
    }

    class Lomo : Filter {
        override fun apply(`in`: Mat): Mat {
            filter(`in`, -1, -1, 1.0, 1.33, 1.33)
            return `in`
        }
    }
}